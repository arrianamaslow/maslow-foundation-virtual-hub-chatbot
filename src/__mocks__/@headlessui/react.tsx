/**
 * @jest-environment jsdom
 */

import { DialogProps } from '@headlessui/react'
import React, { ReactNode } from 'react'

type DialogPanelTestProps = {
  children?: ReactNode
}

interface DialogTestProps extends DialogProps {
  children?: ReactNode
}

const DialogPanel: React.FC<DialogPanelTestProps> = ({ children }) => {
  return <div data-testid='mock-dialog-panel-component'>{children}</div>
}

const Dialog: React.FC<DialogTestProps> = ({ open, children }) => {
  return (
    <div data-testid='mock-dialog-component'>
      {open ? 'DIALOGUEOPENED' : 'DIALOGUECLOSED'}
      {children}
    </div>
  )
}

export { Dialog, DialogPanel }
