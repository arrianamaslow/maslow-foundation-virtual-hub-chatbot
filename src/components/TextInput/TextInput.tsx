import React, { useState } from 'react'
import InputErrorWrapper from '../InputErrorWrapper/InputErrorWrapper'
type TextInputProps = {
  placeholderText?: string
  questionText: string
  validation: (value: string) => string
}

const TextInput: React.FC<TextInputProps> = ({ placeholderText, questionText, validation }) => {
  const [message, setMessage] = useState<string>('')
  const [errorMessage, setErrorMessage] = useState<string>('')

  const runValidation = (inputMessage: string, restrictUpdate: boolean) => {
    const errorMessageReturned = validation(inputMessage)
    if (errorMessageReturned) {
      if (!restrictUpdate) {
        setErrorMessage(errorMessageReturned)
      }
    } else {
      setErrorMessage('')
    }
  }

  return (
    <div className='flex-row md:h-[77px] h-[72px]'>
      <p className='text-left md:text-[1.313rem] text-[17px] font-normal font-lato'>
        {questionText}
      </p>
      <InputErrorWrapper errorMessage={errorMessage}>
        <input
          className={'w-full'}
          data-testid='input-field'
          type='text'
          placeholder={placeholderText}
          onChange={(e) => {
            setMessage(e.currentTarget.value)
            runValidation(e.currentTarget.value, true)
          }}
          onBlur={(e) => {
            runValidation(e.currentTarget.value, false)
          }}
        />
      </InputErrorWrapper>
    </div>
  )
}

export default TextInput
