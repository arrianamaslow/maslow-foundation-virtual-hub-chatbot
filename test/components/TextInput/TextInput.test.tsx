/**
 * @jest-environment jsdom
 */
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

jest.mock('@/components/InputErrorWrapper/InputErrorWrapper', () => {
  const MockWrapper = ({ errorMessage, children }: any) => (
    <div data-testid={errorMessage}>{children}</div>
  )

  return MockWrapper
})

import TextInput from '@/components/TextInput/TextInput'

describe('<TextInput />', () => {
  it('renders correctly', () => {
    const tree = render(
      <TextInput
        placeholderText=''
        questionText=''
        validation={(value: string) => {
          return 'default text'
        }}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('has an input field which is of type text and has placeholder text equal to the value of the placeholder parameter', () => {
    render(
      <TextInput
        placeholderText='placeholder text'
        questionText=''
        validation={(value: string) => {
          return 'default text'
        }}
      />
    )
    const inputField = screen.getByTestId('input-field')
    expect(inputField).toHaveAttribute('type', 'text')
    expect(inputField).toHaveAttribute('placeholder', 'placeholder text')
  })

  it('runs validation when its input field is blured', () => {
    const validation = jest.fn()
    render(<TextInput placeholderText='placeholder text' questionText='' validation={validation} />)
    const inputField = screen.getByTestId('input-field')
    fireEvent.change(inputField, {
      target: { value: 'test string' }
    })
    fireEvent.blur(inputField)
    expect(validation).toHaveBeenCalledWith('test string')
  })

  it('has an input field which displays an error when validation is not met', () => {
    const validation = jest.fn()
    validation.mockReturnValue('invalid-input')
    render(<TextInput placeholderText='placeholder text' questionText='' validation={validation} />)
    const inputField = screen.getByTestId('input-field')
    fireEvent.change(inputField, {
      target: { value: 'test string' }
    })
    fireEvent.blur(inputField)
    const complaintDiv = screen.getByTestId('invalid-input')
    expect(complaintDiv).not.toBeNull()
  })

  it('has an input field which does run validation on change where an error message is displayed', () => {
    const validation = jest.fn()
    validation.mockReturnValue('invalid-input')
    render(<TextInput placeholderText='placeholder text' questionText='' validation={validation} />)
    const inputField = screen.getByTestId('input-field')
    fireEvent.change(inputField, {
      target: { value: 'test string' }
    })
    fireEvent.blur(inputField)
    const complaintDiv = screen.getByTestId('invalid-input')
    expect(complaintDiv).not.toBeNull()
    fireEvent.change(inputField, {
      target: { value: 'test string' }
    })
    expect(validation).toHaveBeenCalledWith('test string')
  })

  it('has an input field which does run validation on change where an error message is not displayed', () => {
    const validation = jest.fn()
    validation.mockReturnValue('invalid-input')
    render(<TextInput placeholderText='placeholder text' questionText='' validation={validation} />)
    const inputField = screen.getByTestId('input-field')
    fireEvent.change(inputField, {
      target: { value: 'test string' }
    })
    expect(validation).toHaveBeenCalledWith('test string')
  })
})
