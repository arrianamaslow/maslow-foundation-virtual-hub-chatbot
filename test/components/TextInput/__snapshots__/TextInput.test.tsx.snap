// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex-row md:h-[77px] h-[72px]"
      >
        <p
          class="text-left md:text-[1.313rem] text-[17px] font-normal font-lato"
        />
        <div
          data-testid=""
        >
          <input
            class="w-full"
            data-testid="input-field"
            placeholder=""
            type="text"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex-row md:h-[77px] h-[72px]"
    >
      <p
        class="text-left md:text-[1.313rem] text-[17px] font-normal font-lato"
      />
      <div
        data-testid=""
      >
        <input
          class="w-full"
          data-testid="input-field"
          placeholder=""
          type="text"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
